server/core/src/main/java/org/infinispan/server/iteration/HashMapIterationManager.java

```diff
@@ -0,0 +1,76 @@
+package org.infinispan.server.iteration;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.CacheStream;
+import org.infinispan.commons.time.TimeService;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.ImmortalCacheEntry;
+import org.infinispan.security.actions.SecurityActions;
+import org.infinispan.stream.impl.local.AbstractLocalCacheStream;
+import org.infinispan.stream.impl.local.LocalCacheStream;
+
+public class HashMapIterationManager extends DefaultIterationManager {
+
+   private static final CacheStream<CacheEntry<Object, Object>> EMPTY_STREAM =
+         new LocalCacheStream<>(new HashMapStreamSupplier(null), false, null);
+
+   public HashMapIterationManager(TimeService timeService) {
+      super(timeService);
+   }
+
+   @Override
+   protected CacheStream<CacheEntry<Object, Object>> baseStream(AdvancedCache cache, IterationInitializationContext ctx) {
+      if (ctx == null) return EMPTY_STREAM;
+      if (!(ctx instanceof HashMapInitializationContext)) throw new IllegalStateException("Illegal context for iteration");
```

Comment:
how about turn ctx into a property of the specialized class, remove it from the method params and avoid this check?

Reply:
Yeah, this was something I was trying at first. I took this approach because the iteration manager is shared between all clients IIUC. So it would need a way to synchronize between retrieving the entry from the cache and invoking start. All this (trying to synchronize or add a ctx) could be avoided if the start method receives the source stream or something similar. Let me think more about this.